fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CPBiRel.java

```diff
@@ -0,0 +1,138 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer;
+
+import com.google.common.collect.Lists;
+import com.starrocks.catalog.ForeignKeyConstraint;
+import com.starrocks.catalog.OlapTable;
+import com.starrocks.common.Pair;
+import com.starrocks.sql.optimizer.operator.logical.LogicalScanOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+// Cardinality-preserving bi-relation, two scan operator form a CPBIRel instance
+// when their underlying OlapTable has {foreign,primary,unique} key constraints and
+// the join equality predicates can match these constraints.
+public class CPBiRel {
+    private final OptExpression lhs;
+    private final OptExpression rhs;
+    private final boolean fromForeignKey;
```

Comment:
Better to  add some comment about what lhs and rhs represents respectively when fromForeignKey is true or false.